### ISSUE TITLE
Expose named fields for non-simd vector types

### ISSUE THREAD

**Issue by termhn** [bitshifter/glam-rs#89]
The motivation is that it's kind of a pain to pay the cost of only accessing fields through function calls when you're not actually using simd. This is coming up for Embark both in WASM and (now) GPU code. Opening this issue to see what your thoughts/concerns are about this.

There's a couple options for implementation... the simplest would be to just make things that can be a plain struct with public named fields, i.e.

```rust
pub struct Vec3 {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}
```

Another option is to use the strategy that `nalgebra` uses which is to create specific "component bag" structs and then impl `Deref` for them for things which make sense. For example:

```rust
struct XYZ {
    x: f32,
    y: f32,
    z: f32,
}

struct RGB {
    r: f32,
    g: f32,
    b: f32,
}
impl Deref<XYZ> for Vec3 {
   //...
}
impl Deref<RGB> for Vec3 {
   //...
}
```

Now you can write both `vec.x` or `vec.r` and you get the same thing as `vec.x()` currently.

**Comment by bitshifter** [comment]
I made the decision early on to keep the API consistent between types that are backed by SIMD and types that aren't as I felt the inconsistency when shifting between types would be more painful than having to deal with accessors everywhere.
 
Aside from consistency between types, there is a consistency with the swizzles that I recently added, e.g. `.xxxx() -> Vec4`, `.xxx() -> Vec3`, `.xx() -> Vec2` and `v.x() -> f32`. Although I think the swizzles being functions may also bother people who are used to common shader languages, not sure if anything can be done about that.

On the WASM front, WASM will get SIMD support eventually in `glam`.

I realise the accessors are a pain and I am not completely opposed to moving away from them long term if that's what users of the library prefer, but it is a big change. This could mean a few possible directions for glam:

* Introducing inconsistencies between different vector types which I feel would be undesirable
* Dropping SIMD entirely and favouring ergonomics over CPU performance if users feel that is more important; or
* Completely separate types for SIMD and scalar - some users have asked for this but I feel it would greatly complicate the library
* Diverging the API when the `scalar-math` feature is used - I think this would be messy internally though

If I was to make a change like this I want to wait and see how the library is used more, both in the game context, particular with Bevy and now for GPU usage. It would be good to see if games benefit much from the SIMD vectors or if their bottlenecks tend to be elsewhere.

As an aside I have in the past experimented with using named fields for vectors and casting them to `__m128` inside functions but there was a significant performance impact from that approach (https://github.com/bitshifter/glam-rs/blob/yolo_align/src/f32/vec4.rs#L23-L28).

The `Deref` suggestion I had not come across before, I will look into that.

**Comment by termhn** [comment]
> Aside from consistency between types, there is a consistency with the swizzles that I recently added, e.g. `.xxxx() -> Vec4`, `.xxx() -> Vec3`, `.xx() -> Vec2` and `v.x() -> f32`. Although I think the swizzles being functions may also bother people who are used to common shader languages, not sure if anything can be done about that.

I'm personally not as bothered by swizzles as functions because they're not as common and the longer the swizzle is the less the () takes up in "syntax overhead". However, I came up with a sorta cursed idea to get the non-function-call syntax (though I'm not sure if this will actually work as it kinda returns a ref to a temporary value?)

The idea is basically to take the `Deref` method outlined above to the next level for swizzles. Basically:

```rust
pub struct SwizzleXX {
    pub xx: Vec2,
}
pub struct SwizzleXXY {
    pub xxy: Vec3,
}
pub struct Swizzle ZYY {
    pub zyy: Vec3,
}

impl Deref<SwizzleXX> for Vec2 { ... }
impl Deref<SwizzleXX> for Vec3 { ... }
impl Deref<SwizzleXX> for Vec4 { ... }

impl Deref<SwizzleXXY> for Vec2 { ... }
impl Deref<SwizzleXXY> for Vec3 { ... }
impl Deref<SwizzleXXY> for Vec4 { ... }

impl Deref<SwizzleZYY> for Vec3 { ... }
impl Deref<SwizzleZYY> for Vec4 { ... }

// ... etc
```

**Comment by repi** [comment]
@termhn how would that look like in code that would be using it? Would it be more ergonomic than say `.xxy()` function call?

**Comment by termhn** [comment]
> @termhn how would that look like in code that would be using it? Would it be more ergonomic than say `.xxy()` function call?

In theory this would allow you to just do `vec.xx`, `vec.xxy`, `vec.zyy` (no function call, looks the same as a glsl/hlsl swizzle)

**Comment by francesco-cattoglio** [comment]
> I made the decision early on to keep the API consistent between types that are backed by SIMD and types that aren't as I felt the inconsistency when shifting between types would be more painful than having to deal with accessors everywhere.
> 
> Aside from consistency between types, there is a consistency with the swizzles that I recently added, e.g. `.xxxx() -> Vec4`, `.xxx() -> Vec3`, `.xx() -> Vec2` and `v.x() -> f32`. Although I think the swizzles being functions may also bother people who are used to common shader languages, not sure if anything can be done about that.
> 
> On the WASM front, WASM will get SIMD support eventually in `glam`.
> 
> I realise the accessors are a pain and I am not completely opposed to moving away from them long term if that's what users of the library prefer, but it is a big change. This could mean a few possible directions for glam:
> 
>     * Introducing inconsistencies between different vector types which I feel would be undesirable
> 
>     * Dropping SIMD entirely and favouring ergonomics over CPU performance if users feel that is more important; or
> 
>     * Completely separate types for SIMD and scalar - some users have asked for this but I feel it would greatly complicate the library
> 
>     * Diverging the API when the `scalar-math` feature is used - I think this would be messy internally though

I am currently using glam and I chose it over the alternatives because the API looks very simple and straightforward, and yet the library offers very good performance. Whatever the final decision will be, my opinion is please keep SIMD support for native and keep the API consistent between SIMD and non-SIMD platforms.

**Comment by not-fl3** [comment]
First of all, thanks for your work on glam! Using it quite a lot and so far it is the best experience over all rust math crates.

However, the best selling point of glam for me is API simplicity and implementation simplicity - it is easy to use and easy to discover features. And with that mindset, I always considered `.x()` calls and restriction on simple vecs destruction is a price to pay and better solution than rolling my own math library/maintaining a fork. 

And with previous `glam` implementation, with different `Vec3` data depending on cargo features - this looked like an inevitable evil. 
But now, with separate Vec3 and Vec3A - it looks doable to get easily accesible vector fields for non-simd users!

And this will make my life and the life of all macroquad users significantly easier, I would not need to answer question on discord about "why all other sturcts in macroquad have public field but vector don't".

**Comment by termhn** [comment]
> Whatever the final decision will be, my opinion is please keep SIMD support for native and keep the API consistent between SIMD and non-SIMD platforms.

I should note that I am not (and I don't think others are) proposing doing away entirely with method-call syntax on non-simd types, so there would still be a consistent API... just that for non-simd, there could be the *addition* of direct-accessible fields :)

**Comment by bitshifter** [comment]
@termhn yeah sorry I neglected to make it clear that my intention was to try your `Deref` suggestion first and foremost, so that should have been at the top of that bullet list and not at the end of my initial comment. The rest of the bullet points are if `Deref` doesn't pan out for some reason.

I expect that adding direct access to `struct` members named `.x` would necessitate removing the function with the same name, which is why I've labelled this as a breaking change. But perhaps I am mistaken about that?

At some point I want to add support for other scalar types like `i32` and `f64` (#70). These most likely would not have a SIMD implementation which I think is more reason to consider direct access.

Anyway it seems like I need to experiment.

**Comment by termhn** [comment]
> I expect that adding direct access to `struct` members named `.x` would necessitate removing the function with the same name, which is why I've labelled this as a breaking change. But perhaps I am mistaken about that?

Nope! https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8c89915615c06086a3612fca4945e9df

**Comment by bitshifter** [comment]
doh, C++ has destroyed my Rust brain.

**Comment by bitshifter** [comment]
I made some progress on this last night. Managed to convert everything to direct access except for Vec3 because I ran out of evening. So far so good. I want to benchmark it and to make sure there's no unexpected perf implications but it seems unlikely that that would be the case.

So hopefully I can push this to master on the weekend. Thank you for the suggestion @termhn! it is much nicer to not have to use accessors, had I realised this was an option I definitely would have done it from the start.

Assuming perf is OK, in particular for reading/writing elements in the SIMD case, the next question is what to do with the accessors. I am inclined to deprecate them now and remove them in a later release. I could add `.x()`, `.y()`, `.z()` and `.w()` to the swizzle traits so that those consistent methods are still there for people who are using swizzles.
